File: 20250831_one_click_paper_page/frontend/src/lib/api/conversion.ts
/**
 * API client for backend conversion endpoints
 */

export interface ConversionJobResponse {
  job_id: string;
  status: string;
  message: string;
}

export interface ConversionStatusResponse {
  job_id: string;
  status: "queued" | "processing" | "completed" | "failed" | "cancelled";
  phase:
    | "queued"
    | "preparing"
    | "analyzing"
    | "converting"
    | "processing"
    | "finalizing"
    | "completed";
  stage: string;
  message: string;
  error?: string;
}

export interface PaperMetadata {
  title?: string;
  authors?: string[];
  abstract?: string;
  keywords?: string[];
  doi?: string;
  arxiv_id?: string;
}

export interface ConversionResult {
  job_id: string;
  status: string;
  output_files: string[];
  metrics: {
    total_conversion_time: number;
    mode_used: string;
    quality_assessment: {
      has_good_text: boolean;
      recommended_mode: string;
      confidence: string;
      avg_chars_per_page: number;
      text_coverage: number;
    };
    model_load_time?: number;
    processing_time?: number;
  };
  markdown_length: number;
  image_count: number;
  html_file: string;
  markdown_file: string;
  metadata?: PaperMetadata;
}

export type ConversionMode = "auto" | "fast" | "quality";

const API_BASE = "http://localhost:8000/api";

export class ConversionAPI {
  /**
   * Upload a file and start conversion
   */
  static async uploadAndConvert(
    file: File,
    template: string,
    mode: ConversionMode = "auto",
    repositoryName?: string
  ): Promise<ConversionJobResponse> {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("template", template);
    formData.append("mode", mode);
    if (repositoryName) {
      formData.append("repository_name", repositoryName);
    }

    const response = await fetch(`${API_BASE}/convert/upload`, {
      method: "POST",
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || "Upload failed");
    }

    return response.json();
  }

  /**
   * Get conversion status
   */
  static async getStatus(jobId: string): Promise<ConversionStatusResponse> {
    const response = await fetch(`${API_BASE}/convert/status/${jobId}`);

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || "Failed to get status");
    }

    return response.json();
  }

  /**
   * Get conversion result
   */
  static async getResult(jobId: string): Promise<ConversionResult> {
    const response = await fetch(`${API_BASE}/convert/result/${jobId}`);

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || "Failed to get result");
    }

    return response.json();
  }

  /**
   * Cancel conversion
   */
  static async cancel(jobId: string): Promise<{ message: string }> {
    const response = await fetch(`${API_BASE}/convert/cancel/${jobId}`, {
      method: "DELETE",
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || "Failed to cancel");
    }

    return response.json();
  }

  /**
   * Poll conversion status until completion
   */
  static async pollStatus(
    jobId: string,
    onProgress?: (status: ConversionStatusResponse) => void,
    intervalMs: number = 2000
  ): Promise<ConversionResult> {
    return new Promise((resolve, reject) => {
      const poll = async () => {
        try {
          const status = await this.getStatus(jobId);

          if (onProgress) {
            onProgress(status);
          }

          if (status.status === "completed") {
            const result = await this.getResult(jobId);
            resolve(result);
          } else if (status.status === "failed") {
            reject(new Error(status.error || "Conversion failed"));
          } else if (status.status === "cancelled") {
            reject(new Error("Conversion was cancelled"));
          } else {
            // Continue polling
            setTimeout(poll, intervalMs);
          }
        } catch (error) {
          reject(error);
        }
      };

      poll();
    });
  }
}

/**
 * React hook for conversion operations
 */
export function useConversion() {
  const [isConverting, setIsConverting] = useState(false);
  const [phase, setPhase] = useState<string>("queued");
  const [stage, setStage] = useState("");
  const [error, setError] = useState<string | null>(null);
  const [result, setResult] = useState<ConversionResult | null>(null);
  const [jobId, setJobId] = useState<string | null>(null);

  const startConversion = async (
    file: File,
    template: string,
    mode: ConversionMode = "auto",
    repositoryName?: string
  ) => {
    setIsConverting(true);
    setPhase("preparing");
    setStage("Uploading file...");
    setError(null);
    setResult(null);

    try {
      // Start conversion
      const job = await ConversionAPI.uploadAndConvert(
        file,
        template,
        mode,
        repositoryName
      );
      setJobId(job.job_id);

      // Poll for completion
      const result = await ConversionAPI.pollStatus(job.job_id, status => {
        setPhase(status.phase);
        setStage(status.message);
      });

      setResult(result);
      setStage("Conversion completed!");
      setPhase("completed");
    } catch (err) {
      setError(err instanceof Error ? err.message : "Conversion failed");
    } finally {
      setIsConverting(false);
    }
  };

  const reset = () => {
    setIsConverting(false);
    setPhase("queued");
    setStage("");
    setError(null);
    setResult(null);
    setJobId(null);
  };

  return {
    isConverting,
    phase,
    stage,
    error,
    result,
    jobId,
    startConversion,
    reset,
  };
}

// Import React for the hook
import { useState } from "react";

File: 20250831_one_click_paper_page/backend/models/conversion.py
"""
Pydantic models for conversion API endpoints.
"""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class ConversionMode(str, Enum):
    """Conversion mode options."""
    AUTO = "auto"      # Smart mode: try fast, fallback to quality if needed
    FAST = "fast"      # Fast mode: disable OCR (~40 seconds)
    QUALITY = "quality"  # Quality mode: full OCR (~6 minutes)


class ConversionStatus(str, Enum):
    """Conversion job status."""
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class FileUploadRequest(BaseModel):
    """Request model for file upload."""
    template: str = Field(..., description="Template to use for the website")
    mode: ConversionMode = Field(ConversionMode.AUTO, description="Conversion mode")
    repository_name: str | None = Field(None, description="Custom repository name")


class ConversionJobResponse(BaseModel):
    """Response model for conversion job creation."""
    job_id: str = Field(..., description="Unique job identifier")
    status: ConversionStatus = Field(..., description="Current job status")
    message: str = Field(..., description="Human-readable status message")


class ConversionPhase(str, Enum):
    """Conversion phases."""
    QUEUED = "queued"
    PREPARING = "preparing"
    ANALYZING = "analyzing"
    CONVERTING = "converting"
    PROCESSING = "processing"
    FINALIZING = "finalizing"
    COMPLETED = "completed"


class ConversionStatusResponse(BaseModel):
    """Response model for conversion status check."""
    job_id: str = Field(..., description="Job identifier")
    status: ConversionStatus = Field(..., description="Current job status")
    phase: ConversionPhase = Field(..., description="Current conversion phase")
    stage: str = Field(..., description="Current processing stage")
    message: str = Field(..., description="Human-readable status message")
    progress: int = Field(..., description="Conversion progress percentage (0-100)")
    error: str | None = Field(None, description="Error message if failed")


class QualityAssessment(BaseModel):
    """PDF quality assessment results."""
    has_good_text: bool = Field(..., description="Whether PDF has good text extraction")
    recommended_mode: ConversionMode = Field(
        ..., description="Recommended conversion mode"
    )
    confidence: str = Field(..., description="Confidence level: low, medium, high")
    avg_chars_per_page: float = Field(..., description="Average characters per page")
    text_coverage: float = Field(..., description="Percentage of pages with text")


class ConversionMetrics(BaseModel):
    """Conversion performance metrics."""
    total_conversion_time: float = Field(
        ..., description="Total conversion time in seconds"
    )
    mode_used: ConversionMode = Field(..., description="Actual conversion mode used")
    quality_assessment: QualityAssessment = Field(
        ..., description="PDF quality assessment"
    )
    model_load_time: float | None = Field(
        None, description="Model loading time in seconds"
    )
    processing_time: float | None = Field(
        None, description="Processing time in seconds"
    )


class PaperMetadata(BaseModel):
    """Extracted paper metadata."""
    title: str | None = None
    authors: list[str] = Field(default_factory=list)
    abstract: str | None = None
    keywords: list[str] = Field(default_factory=list)
    doi: str | None = None
    arxiv_id: str | None = None


class ConversionResult(BaseModel):
    """Complete conversion result."""
    job_id: str = Field(..., description="Job identifier")
    status: ConversionStatus = Field(..., description="Final job status")
    success: bool = Field(..., description="Whether conversion was successful")
    output_dir: str = Field(..., description="Output directory path")
    output_files: list[str] = Field(..., description="List of generated output files")
    metrics: ConversionMetrics = Field(..., description="Performance metrics")
    markdown_length: int = Field(..., description="Length of extracted markdown")
    image_count: int = Field(..., description="Number of extracted images")
    html_file: str = Field(..., description="Path to generated HTML file")
    markdown_file: str = Field(..., description="Path to generated markdown file")
    metadata: PaperMetadata | None = Field(None, description="Extracted paper metadata")


class ConversionError(BaseModel):
    """Error response model."""
    job_id: str = Field(..., description="Job identifier")
    error_type: str = Field(..., description="Type of error")
    message: str = Field(..., description="Error message")
    details: dict[str, Any] | None = Field(None, description="Additional error details")


class WebSocketMessage(BaseModel):
    """WebSocket message for real-time updates."""
    job_id: str = Field(..., description="Job identifier")
    type: str = Field(
        ..., description="Message type: progress, status, error, complete"
    )
    data: dict[str, Any] = Field(..., description="Message data")
    timestamp: float = Field(..., description="Unix timestamp")
